### ISSUE TITLE
Proposal: Updated implementation for the In-toto spec

### ISSUE THREAD

**Issue by Alos** [in-toto/in-toto-java#23]
cc @SantiagoTorres  @tomhennen

Hello! :)

As described in this [request](https://github.com/in-toto/attestation/issues/56) we want to contribute a new version of the Java client. We would like to implement something like the following:

<img width="1220" alt="Screen Shot 2021-08-16 at 11 06 16 AM" src="https://user-images.githubusercontent.com/25460/129609324-5463902f-393c-4f13-8414-ce180fd6b073.png">

The above diagram contains the models used by the Java library to represent the various entities required from the different specs (in-toto, SLSA, DSSE) to generate build provenance. 

A few notes on the proposed classes:

* The Recipe entity has additional properties: arguments and environment that are classified as generic objects in the spec. For this reason it would be expected that builders using this library would extend from the default Recipe and create their own custom class.
* The entities are contained within packages with naming that matches the origin of the spec, i.e. Provenance from SLSA, Envelope from DSEE, etc.  It would be expected that the implementation for the different specs resides in the corresponding GitHub project.

### Structural changes

* Create a new branch to contain the new version for the proposal
* Add a legacy package that contains the existing implementation as is.
* Clean up and update the pom.xml

### API changes

* Remove signing aspect in the library
* Create helper classes to convert to JSON

### General changes

* Replace gson with Jackson and annotations
* Only use lowercase letter with no underscores in package names
* Do not have multiple public nested classes

Currently the library has a POJO (the Link) that not only holds information about the entity it represents but also is able to transform it self to JSON and sign it self. 

A proposed change would be for serialization to JSON to happen in a helper class using Jackson. Something like:

```java
Statement myStatement = new Statement()
    .setSubject(…)
    .setPedicateType(…)
    .setPredicate(…);
String jsonString = IntotoHelper.toJson(myStatement);
```

For this to happen, the new POJOs will have [Jackson](https://github.com/FasterXML/jackson-core) annotations where required. The _IntotoHelper_ class would instantiate Jackson and configure it so that it outputs the correct JSON format.
Currently POJOs that want to be signable extend from Metablock. Metablock exposes the following method:

```java
sign(Key privateKey)
```

The proposed change is to have an external class (probably an implementation of an interface in the DSSE repo) perform the signing. Allowing users to either provide a key to sign or sign the payloads using their own mechanism (i.e Cloud KMS). 

## Proposed changes in: io.github.dsse
Create a new java library that contains Envelope, Signature and helper classes to implement signing. This library should have the implementations to sign with a given Key (similar to the current implementation in the Java lib) but also allow the capability for users to define their own Signers that would allow the dsse library to correctly create the Envelopes. 

## Proposed changes in: io.github.slsa
Create a new Java library that contains implementation of the Provenance spec. This would contain the models to represent a Provenance as well as helper classes to serialize this into JSON. 

## Questions

* How and where to run CI/CD tests? Github actions? Cloud Build?
* Should we target JDK 11 or JDK 17?
* Please let me know if something does not make sense, if I made a mistake or would like to see a different implementation.

**Comment by SantiagoTorres** [comment]
Wow, thank you so much for the detailed plan! From a quick, cursory read, this looks excellent, although I think we could discuss some of the specifics on a one-by-one basis :)

The class diagram though looks fantastic, I quite like it :)

**Comment by SantiagoTorres** [comment]
I'll have more comments as I think things through...

**Comment by SantiagoTorres** [comment]
I think we should just give it a go. We can discuss minutiae once we see PRs

**Comment by Alos** [comment]
BTW, I do have a section with questions, specifically, I'm interested in finding out if there is an exiting way of running tests or a  CI/CD set up already done?

**Comment by adityasaky** [comment]
Not on this repository. We use GitHub Actions on the other repositories for the most part.

**Comment by Alos** [comment]
I see. It looks like the existing implementation has issues passing the tests. I'll try to fix these too.

**Comment by adityasaky** [comment]
Thank you!

**Comment by Alos** [comment]
Ok I have all the core functionality of the stuff inside intoto package:

<img width="221" alt="Screen Shot 2021-09-10 at 12 58 18 PM" src="https://user-images.githubusercontent.com/25460/132910981-10a8137a-8a78-4a1b-ad72-f0ed60059867.png">


Now the question is: 

Do I add implementations for DSSE in the DSSE repository, and implementations for the SLSA Provenance in the SLSA repository? or do I bundle all the functionality into this library? or do we create the bundle in the SLSA repo? How are other languages doing it?

##Implementing in their own repo

Pros:
* We enforce the idea that DSSE and SLSA are not part of in-toto and instead their own projects
* We keep the existing library simple
* Users can import only what they need

Cons:
* It would be awkward for users to have to pull from multiple repos to create say a Provenance, make an attestation and sign it into a DSSE.
* Users might want to just import a single library that has everything.
* Have to get permissions from other projects

##Implementing it in a single repo

Pros:
* Users just import a single library that can do it all
* Easy to keep the UX consistent
* Already have access to this repo

Cons:
* Dirties the idea of the 3 different projects
* Unsure if a complete library should remain here or needs to be moved to another repo

Bonus question:
* How do we get a proper CI/CD going?

**Comment by SantiagoTorres** [comment]
So what we did with the other implementations was to start in-repo and then strip out the module when needed (to be still included as a build time dependency to abstract out the interface). I wonder if this is a reasonable path here as well

**Comment by SantiagoTorres** [comment]
Also, let me help you out to set up the CI/CD this weekend!

**Comment by adityasaky** [comment]
Closed by #24
